### PR TITLE
Introduce the Attribute criterion

### LINKED ISSUE #108: Implement attribute criterion
Related to #6 🚀

### PR DESCRIPTION
**Pull request by Kerollmops**
This PR introduces the Attribute criterion along with the internal `words-level-positions-docids` database required. This criterion is designed to be performant when the number of candidates is small or big. It adopts the right algorithm for both situations. When it is small, a linear algorithm is used, and when it is big it uses the `words-level-positions-docids` database to do unions and intersections.

Closes #108.

## What must be done before we can merge

 - [x] Implement the set-theory-based algorithm for when the number of candidates is big.
 - [x] A request from @LegendreM is to create only word positions level groups on deterministic bounds. This means that instead of doing group levels of runs of four existing positions, we should instead do group levels of runs of potentially existing positions. If there are words at positions 0, 1, 3, and 4, we should do two groups (0..=3 and 4..=7) instead of one group (0..=4).

### CODE REVIEW

**Comment by LegendreM** [comment]
@Kerollmops can you review this PR,

@irevoire could you give an external eye to this PR,

Thanks!

**Review by Kerollmops — COMMENTED** [review]
Here is a partial review, I need to check the Attribute criterion itself.

**Review by irevoire — APPROVED** [review]

File: milli/src/search/criteria/proximity.rs [hunk 1]
```diff
@@ -8,48 +8,29 @@ use log::debug;
...
+    pub fn new(ctx: &'t dyn Context<'t>, parent: Box<dyn Criterion + 't>) -> Self {
         Proximity {
             ctx,
-            query_tree: None,
+            state: None,
             proximity: 0,
-            candidates: Candidates::default(),
             bucket_candidates: RoaringBitmap::new(),
-            parent: Some(parent),
+            parent: parent,
```
Comment (Kerollmops) [inline milli/src/search/criteria/proximity.rs:33]:
```suggestion
            parent,
```

File: milli/src/search/criteria/attribute.rs [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+    in_prefix_cache: bool,
+    inner_next: Option<(u32, u32, RoaringBitmap)>,
+    current_interval: Option<(u32, u32)>,
+}
+
+impl<'t, 'q> WordLevelIterator<'t, 'q> {
+    fn new(ctx: &'t dyn Context<'t>, word: Cow<'q, str>, in_prefix_cache: bool) -> heed::Result<Option<Self>> {
+        match ctx.word_position_last_level(&word, in_prefix_cache)? {
+            Some(level) =>  {
+                let interval_size = 4u32.pow(Into::<u8>::into(level.clone()) as u32);
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:155]: We should define this 4 in a const at the top of this file :)

File: milli/src/search/criteria/attribute.rs [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+        Ok(accumulated)
+    }
+
+    /// return the next meta-interval created from inner WordLevelIterators,
+    /// and from eventual chainned QueryLevelIterator.
+    fn next(&mut self, allowed_candidates: &RoaringBitmap, tree_level: TreeLevel) -> heed::Result<Option<(u32, u32, RoaringBitmap)>> {
+        let parent_result = match self.parent.as_mut() {
+            Some(parent) => {
+                Some(parent.next(allowed_candidates, tree_level)?)
+            },
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:311]:
```suggestion
            Some(parent) => Some(parent.next(allowed_candidates, tree_level)?),
```

File: milli/src/search/criteria/attribute.rs [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+                        },
+                        (Some((_, _, inner)), None) => {
+                            inner.is_disjoint(allowed_candidates)
+                        },
+                        (None, Some((_, _, parent))) => {
+                            parent.is_empty()
+                        },
+                        (None, None) => true,
+                    }
+                }).count();
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:331]: Could you please reduce this code block by maybe using multiple variables and computing intermediate values, first?

File: milli/src/search/criteria/attribute.rs [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+                        self.accumulator = vec![None];
+                        Ok(None)
+                    },
+                }
+            }
+        }
+    }
+}
+
+struct Branch<'t, 'q> {
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:365]: Could you please add a little bit of documentation on this struct too, please?

File: milli/src/search/criteria/attribute.rs [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+                None => {
+                    // the branch seems to be invalid, so we skip it.
+                    branch_positions.clear();
+                    break;
+                },
+            }
+        }
+        // QueryLevelIterator need to be sorted by level and folded in descending order.
+        branch_positions.sort_unstable_by_key(|qli| qli.level);
+        let folded_query_level_iterators =  branch_positions
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:458]:
```suggestion
        let folded_query_level_iterators = branch_positions
```

File: milli/src/search/criteria/attribute.rs [hunk 7]
```diff
@@ -0,0 +1,715 @@
...
+                Some(fold) => {
+                    qli.parent(fold);
+                    Some(qli)
+                },
+                None => Some(qli),
+        });
+
+        if let Some(mut folded_query_level_iterators) = folded_query_level_iterators {
+            let tree_level = folded_query_level_iterators.level;
+            let last_result  = folded_query_level_iterators.next(allowed_candidates, tree_level)?;
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:470]:
```suggestion
            let last_result = folded_query_level_iterators.next(allowed_candidates, tree_level)?;
```

File: milli/src/search/criteria/attribute.rs [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+    let lowest_level = TreeLevel::min_value();
+    let mut final_candidates: Option<(u32, RoaringBitmap)> = None;
+    let mut allowed_candidates = allowed_candidates.clone();
+
+    while let Some(mut branch) = branches_heap.peek_mut() {
+        let is_lowest_level = branch.tree_level == lowest_level;
+        let branch_rank = branch.compute_rank();
+        // if current is worst than best we break to return
+        // candidates that correspond to the best rank
+        if let Some((best_rank, _)) = final_candidates { if branch_rank > best_rank { break; } }
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:503]:
```suggestion
        if let Some((best_rank, _)) = final_candidates {
            if branch_rank > best_rank { break }
        }
```

File: milli/src/search/criteria/attribute.rs [hunk 9]
```diff
@@ -0,0 +1,715 @@
...
+            // we have candidates, lets dig deeper in levels.
+            branch.dig(ctx)?;
+            if !branch.next(&allowed_candidates)? { PeekMut::pop(branch); }
+        }
+
+    }
+
+    Ok(final_candidates.map(|(_rank, candidates)| {
+        candidates
+    }))
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:536]:
```suggestion
    Ok(final_candidates.map(|(_rank, candidates)| candidates))
```

File: milli/src/search/criteria/attribute.rs [hunk 10]
```diff
@@ -0,0 +1,715 @@
...
+            let mut branch_rank = Vec::with_capacity(branch_len);
+            for derivates in branch {
+                let mut position = None;
+                for Query { prefix, kind } in derivates {
+                    // find the best position of the current word in the document.
+                    let current_position = match kind {
+                        QueryKind::Exact { word, .. } => {
+                            if *prefix {
+                                word_derivations(word, true, 0, &words_positions)
+                                .flat_map(|positions| positions.iter().next()).min()
```
Comment (Kerollmops) [inline milli/src/search/criteria/attribute.rs:559]:
```suggestion
                                    .flat_map(|positions| positions.iter().next()).min()
```

**Review by Kerollmops — COMMENTED** [review]
Thank you very much, I think we can merge now!

**Pull request merged**